### PR TITLE
Update "Ways to run a Q# program"

### LINKED ISSUE #190: Document passing parameters and getting outputs from Q#
Sections "Q# with host programs" and "Q# Jupyter Notebooks" have sections on passing parameters, but they only describe very simple scenarios such as passing an integer. Some of the things a person who's trying to do something more interesting will struggle with are:

* Passing arrays - `QArray` is not documented, and only occurs in API documentation.
* Passing callables as arguments.
* Passing qubits as arguments from classical code (not possible, but needs documentation on this).
* Receiving results from operations, the nuances of tuples and arrays in the results.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 206e9bfe-7e73-7f7d-5243-95978ec0d613
* Version Independent ID: fea0833b-00dd-d61a-f197-115050e1b65f
* Content: [Ways to run a Q# program - Azure Quantum ](https://docs.microsoft.com/en-us/azure/quantum/user-guide/host-programs?tabs=tabid-python)
* Content Source: [articles/user-guide/host-programs.md](https://github.com/MicrosoftDocs/quantum-docs/blob/main/articles/user-guide/host-programs.md)
* Service: **azure-quantum**
* Sub-service: **qsharp-guide**
* GitHub Login: @gillenhaalb
* Microsoft Alias: **a-gibec**

### PR DESCRIPTION
**Pull request by SoniaLopezBravo**
This PR contains the following updates:

- Passing arguments using host programs
- Minor edits
- Enhancement of code snippets

This closes #190

### CODE REVIEW

**Review by Bradben — APPROVED** [review]
Looks good, thanks!

File: articles/user-guide/host-programs.md [hunk 1]
```diff
@@ -273,24 +279,22 @@ To work with different host programs, there are no changes required to a `*.qs`
...
+    open Microsoft.Quantum.Arrays;        // for ForEach
 
     operation MeasureSuperposition() : Result {
-        use q = Qubit();
-        H(q);
-        return MResetZ(q);
-
+        use q = Qubit();   // allocates qubit for use (automatically in |0>)
+        H(q);                // puts qubit in superposition of |0> and |1>
+        return MResetZ(q);   // measures qubit, returns result (and resets it to |0> before deallocation)
```
Comment (dphansen) [inline articles/user-guide/host-programs.md:291]:
```suggestion
        use q = Qubit();     // allocates qubit for use (automatically in |0>)
        H(q);                // puts qubit in superposition of |0> and |1>
        return MResetZ(q);   // measures qubit, returns result (and resets it to |0> before deallocation)
```

File: articles/user-guide/host-programs.md [hunk 2]
```diff
@@ -170,15 +168,23 @@ Additionally, it's no problem if your callable includes [documentation comments]
 
 So far, we've only considered an operation that takes no inputs.
 Suppose we wanted to perform a similar operation, but on multiple qubits---the number of which is provided as an argument.
-Such an operation could be written as
+Such an operation could be written as:
```
Comment (dphansen) [inline articles/user-guide/host-programs.md:171]:
```suggestion
Such an operation can be written as:
```

File: articles/user-guide/host-programs.md [hunk 3]
```diff
@@ -140,22 +140,20 @@ However, the [`dotnet run` command](/dotnet/core/tools/dotnet-run) accepts numer
...
 
     @EntryPoint()
     operation MeasureSuperposition() : Result {
-        use q = Qubit();
-        H(q);
-        return MResetZ(q);
-
+        use q = Qubit();   // allocates qubit for use (automatically in |0>)
+        H(q);                // puts qubit in superposition of |0> and |1>
+        return MResetZ(q);   // measures qubit, returns result (and resets it to |0> before deallocation)
```
Comment (dphansen) [inline articles/user-guide/host-programs.md:156]:
```suggestion
        use q = Qubit();     // allocates qubit for use (automatically in |0>)
        H(q);                // puts qubit in superposition of |0> and |1>
        return MResetZ(q);   // measures qubit, returns result (and resets it to |0> before deallocation)
```

File: articles/user-guide/host-programs.md [hunk 4]
```diff
@@ -116,7 +116,7 @@ More specifically, the differences revolve around:
 - Specifying the target machine on which to run it
 - How any results are returned
 
-First, we discuss how this is done with the Q# standalone application from the command prompt, and then proceed to using Python and C# host programs.
+In the following sections we will discuss how this is done with the Q# standalone application from the command prompt, and then proceed to using Python and C# host programs.
 We reserve the standalone application of Q# Jupyter Notebooks for last, because unlike the first three, it's primary functionality does not center around a local Q# file.
 
 > [!NOTE]
```
Comment (dphansen) [inline articles/user-guide/host-programs.md:122]: We shouldn't use "we" in our docs. Would you be able to rewrite these two sentences (and the note in line 123) to not use that)? Also, if you could check the rest of the article for that, that'd be great. Thank you.

File: articles/user-guide/host-programs.md [hunk 5]
```diff
@@ -140,22 +140,20 @@ However, the [`dotnet run` command](/dotnet/core/tools/dotnet-run) accepts numer
 ### Add entry point to Q# file
 
 Most Q# files will contain more than one callable, so naturally we need to let the compiler know *which* callable to run when we provide the `dotnet run` command.
-This is done with a simple change to the Q# file itself:
-    - add a line with `@EntryPoint()` directly preceding the callable.
+This is done with a simple change to the Q# file itself; you need to add a line with `@EntryPoint()` directly preceding the callable.
 
-Our file from above would therefore become
+Our file from above would therefore become:
```
Comment (dphansen) [inline articles/user-guide/host-programs.md:145]: Who's file? We should avoid terms like "our" and "we" in the docs.
```suggestion
The file from above would therefore become:
```

File: articles/user-guide/host-programs.md [hunk 6]
```diff
@@ -367,7 +371,14 @@ Multiple qubits:
 [0, 1, 1, 1]
 {'CNOT': 0, 'QubitClifford': 4, 'R': 0, 'Measure': 4, 'T': 0, 'Depth': 0, 'Width': 4, 'BorrowedWidth': 0}
 ```
+Passing arrays in a similar manner is also possible. You can see an example of that in the [samples](https://github.com/microsoft/Quantum/blob/main/samples/algorithms/reversible-logic-synthesis/host.py#L15).
```
Comment (dphansen) [inline articles/user-guide/host-programs.md:374]:
```suggestion
Passing arrays in a similar manner is also possible. You can see an example of that in the [Reversible Logic Synthesis sample](https://github.com/microsoft/Quantum/blob/main/samples/algorithms/reversible-logic-synthesis/host.py#L15). 
```

File: articles/user-guide/host-programs.md [hunk 7]
```diff
@@ -367,7 +371,14 @@ Multiple qubits:
...
+
+Passing qubits as arguments from classical code is not possible. Any logic that relates to Q# types like `Qubit` should live in your Q# code. If you want your Python code to specify the number of qubits, you could have something like `nQubits : Int` parameter to your Q# operation. Your Python code could pass this as a integer and then your Q# code could allocate the array of the appropriate number of qubits.
 
+For the `Pauli` and `Result` types, there are actually Python enums defined such that you could pass those values directly if you want to. See [qsharp.Pauli](https://docs.microsoft.com/python/qsharp-core/qsharp.pauli?azure-portal=true!) and [qsharp.Result](https://docs.microsoft.com/python/qsharp-core/qsharp.result?azure-portal=true!).
+
+Passing a previously defined callable such as an operation as argument is possible in Python host programs. You can see an example of passing callables to Q# in the [samples](https://github.com/microsoft/Quantum/blob/main/samples/interoperability/python/tomography-sample.ipynb).
```
Comment (dphansen) [inline articles/user-guide/host-programs.md:380]:
```suggestion
Passing a previously defined callable such as an operation as argument is possible in Python host programs. You can see an example of passing callables to Q# in the [Quantum Process Tomography with Q# and Python sample](https://github.com/microsoft/Quantum/blob/main/samples/interoperability/python/tomography-sample.ipynb).
```

File: articles/user-guide/host-programs.md [hunk 8]
```diff
@@ -623,7 +655,7 @@ BorrowedWidth   0
 Q# Jupyter Notebooks make use of the IQ# kernel, which allows you to define, compile, and run Q# callables in a single notebook---all alongside instructions, notes, and other content.
 This means that while it is possible to import and use the contents of `*.qs` Q# files, they are not necessary in the run model.
 
-Here, we will detail how to run the Q# operations defined above, but a more broad introduction to using Q# Jupyter Notebooks is provided at [Intro to Q# and Jupyter Notebooks](https://github.com/microsoft/Quantum/blob/main/samples/getting-started/intro-to-iqsharp/Notebook.ipynb).
+Here, we will detail how to run the Q# operations defined above, but a more broad introduction to using Q# Jupyter Notebooks is provided at [Intro to Q# with Jupyter Notebooks](xref:microsoft.quantum.install-qdk.overview.jupyter).
```
Comment (dphansen) [inline articles/user-guide/host-programs.md:658]:
```suggestion
Here, we will detail how to run the Q# operations defined above, but a more broad introduction to using Q# Jupyter Notebooks is provided at [Introduction to Q# with Jupyter Notebooks](xref:microsoft.quantum.install-qdk.overview.jupyter).
```

**Review by dphansen — CHANGES_REQUESTED** [review]
I have a few minor suggestions for changes.

File: articles/user-guide/host-programs.md (hunk 4 above)
Comment (SoniaLopezBravo) [reply articles/user-guide/host-programs.md:122]: Ok, I will. But I think the use of "we" is widespread in the docs :(

**Review by dphansen — APPROVED** [review]
Looks good.

**Pull request merged**